language/vm/vm_runtime/src/interpreter.rs

```diff
@@ -0,0 +1,1208 @@
+                // FIXME: Once we add in memory ops, we will need to pass in the current memory size
+                // to this function.
+                self.gas_meter.calculate_and_consume(
+                    instruction,
+                    InterpreterForGasCost::new(&self.operand_stack, &self.module_cache, frame),
+                    AbstractMemorySize::new(1),
+                )?;
+                frame.pc += 1;
+
+                match instruction {
+                    Bytecode::Pop => {
+                        self.operand_stack.pop()?;
+                    }
+                    Bytecode::Ret => {
+                        return Ok(ExitCode::Return);
+                    }
+                    Bytecode::BrTrue(offset) => {
+                        if self.operand_stack.pop_as::<bool>()? {
+                            frame.pc = *offset;
+                            break;
+                        }
+                    }
+                    Bytecode::BrFalse(offset) => {
+                        if !self.operand_stack.pop_as::<bool>()? {
+                            frame.pc = *offset;
+                            break;
+                        }
+                    }
+                    Bytecode::Branch(offset) => {
+                        frame.pc = *offset;
+                        break;
+                    }
+                    Bytecode::LdConst(int_const) => {
+                        self.operand_stack.push(Value::u64(*int_const))?;
+                    }
+                    Bytecode::LdAddr(idx) => {
+                        self.operand_stack
+                            .push(Value::address(*frame.module().address_at(*idx)))?;
+                    }
+                    Bytecode::LdStr(idx) => {
+                        self.operand_stack
+                            .push(Value::string(frame.module().user_string_at(*idx).into()))?;
+                    }
+                    Bytecode::LdByteArray(idx) => {
+                        self.operand_stack.push(Value::byte_array(
+                            frame.module().byte_array_at(*idx).clone(),
+                        ))?;
+                    }
+                    Bytecode::LdTrue => {
+                        self.operand_stack.push(Value::bool(true))?;
+                    }
+                    Bytecode::LdFalse => {
+                        self.operand_stack.push(Value::bool(false))?;
+                    }
+                    Bytecode::CopyLoc(idx) => {
+                        self.operand_stack.push(frame.copy_loc(*idx)?)?;
+                    }
+                    Bytecode::MoveLoc(idx) => {
+                        self.operand_stack.push(frame.move_loc(*idx)?)?;
+                    }
+                    Bytecode::StLoc(idx) => {
+                        frame.store_loc(*idx, self.operand_stack.pop()?)?;
+                    }
+                    Bytecode::Call(idx, _) => {
+                        return Ok(ExitCode::Call(*idx));
```

Comment:
Had a look briefly and it should be fine. I'm expecting a bit of rebasing, but since we have stacks and frames in both the old and new versions, the additional logic shouldn't need to be changed much.